src/Workspaces/CoreTest/WorkspaceServiceTests/DirectMemoryAccessStreamReaderTests.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+using static Microsoft.CodeAnalysis.Host.TemporaryStorageServiceFactory;
+
+namespace Microsoft.CodeAnalysis.UnitTests.WorkspaceServiceTests
+{
+    [Trait(Traits.Feature, Traits.Features.Workspace)]
+    public sealed class DirectMemoryAccessStreamReaderTests
+    {
+        [Fact]
+        public void PeakRead()
+        {
+            using var _ = CreateReader("text", out var reader);
+            Assert.Equal('t', reader.Read());
+            Assert.Equal('e', reader.Peek());
+            Assert.Equal('e', reader.Read());
+            Assert.Equal('x', reader.Read());
+            Assert.Equal('t', reader.Read());
+            Assert.Equal(-1, reader.Peek());
+            Assert.Equal(-1, reader.Read());
+        }
+
+        [Theory]
+        [InlineData(5, 0, 5, "bcdef", 5, 'g')]
+        [InlineData(5, 0, 2, "bc\0\0\0", 2, 'd')]
+        [InlineData(10, 1, 6, "\0bcdefg\0\0\0", 6, 'h')]
+        [InlineData(10, 1, 7, "\0bcdefgh\0\0", 7, -1)]
+        [InlineData(10, 1, 9, "\0bcdefgh\0\0", 7, -1)]
+        [InlineData(10, 3, 7, "\0\0\0bcdefgh", 7, -1)]
+        [InlineData(10, 3, 0, "\0\0\0\0\0\0\0\0\0\0", 0, 'b')]
+        [InlineData(10, 10, 0, "\0\0\0\0\0\0\0\0\0\0", 0, 'b')]
+        [InlineData(0, 0, 0, "", 0, 'b')]
+        public void ReadToArray(int bufferLength, int index, int count, string expected, int expectedResult, int expectedPeek)
```

Comment:
💡 In a situation like this, I would typically make two changes:

1. I would remove all of the "expected" parameters. This causes the test to be parameterized only on the inputs, and if the outputs ever change the test signature itself doesn't change and therefore the CI history preserves the long-term test statistics.
2. Instead of hard-coding the expected outputs, I would create a reference stream instance alongside the one we are testing (e.g. `MemoryStream`). Then instead of verifying specific values, I would assert that both streams produce the same result from the same API calls.

Reply:
@sharwell  Ah, I see what you mean. Comparing the behavior to a built-in `TextReader` like `StringReader` or `StreamReader`? The expected outputs shouldn't ever change and I think it's useful to see the expected outputs in the test code like this so that we can easily see what sorts of interesting cases are tested, including their outputs, and whether there's an interesting case that we're missing.

Reply:
I think the point here is that the outputs aren't relevant (and on their own don't tell us if this is correct or not).  What we want to show is that this type behaves identically to any other stream-reader type in terms of semantics.  so as long as they all agree on the behavior, then this is a suitable type to use.

Reply:
@CyrusNajmabadi Why are they not relevant? I think the outputs are really useful so that you can see 1) whether the behavior is correct 2) what sorts of interesting case we're testing, including the various outputs we're testing. If it was only compared to a built-in `TextReader`, you wouldn't be able to see what sorts of different kinds of outputs we're testing.

Reply:
Also, the inputs are chosen partially based on the different outputs we want to get and test. If you don't see the output, you can't compare the inputs to the outputs easily and see why some of these particular inputs are chosen.

Reply:
Would it be OK if I added an additional test that compares to an existing reader without getting rid of the tests that use expected outputs?

Reply:
> @CyrusNajmabadi Why are they not relevant? I think the outputs are really useful so that you can see 1) whether the behavior is correct

Because i woudl have to figure out what correctness means.  Which means i could make a mistake there.  I trust the battle-tested shipping impls of this APIs more than me.  So i'd rather compare the results to *them* rather to me :)